Source/Core/UICommon/NetPlayIndex.cpp

```diff
@@ -0,0 +1,297 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "UICommon/NetPlayIndex.h"
+
+#include <picojson/picojson.h>
+
+#include "Common/HttpRequest.h"
+#include "Common/Thread.h"
+#include "Common/Version.h"
+
+#include "Core/Config/NetplaySettings.h"
+
+NetPlayIndex::NetPlayIndex()
+{
+}
+
+NetPlayIndex::~NetPlayIndex()
+{
+  if (!m_secret.empty())
+    Remove();
+}
+
+static std::optional<picojson::value> ParseResponse(std::optional<std::vector<u8>> response)
+{
+  std::string response_string(reinterpret_cast<char*>(response->data()), response->size());
+
+  picojson::value json;
+
+  auto error = picojson::parse(json, response_string);
+
+  if (!error.empty())
+    return {};
+
+  return json;
+}
+
+std::optional<std::vector<NetPlaySession>>
+NetPlayIndex::List(const std::map<std::string, std::string>& filters)
+{
+  std::vector<NetPlaySession> sessions;
+
+  Common::HttpRequest request;
+
+  std::string list_url = Config::Get(Config::NETPLAY_INDEX_URL) + "/v0/list";
+
+  if (!filters.empty())
+  {
+    list_url += "?";
+    for (const auto& filter : filters)
+    {
+      list_url += filter.first + "=" + request.EscapeComponent(filter.second) + "&";
+    }
+    list_url = list_url.substr(0, list_url.size() - 1);
+  }
+
+  auto response = request.Get(list_url);
+
+  if (!response)
+  {
+    m_last_error = "NO_RESPONSE";
```

Comment:
These response checks seem to be duplicated in a few places.

Reply:
I don't think those can be deduplicated in a reasonable manner.

Reply:
It looks like a few could be replaced by a function that takes a url, performs a request, parses json and returns the error/status.